[GPU Process] [FormControls] Make it possible to draw a viewless NSCell

shallawa opened the pull request:
#### b8b3d9929d7dafe50bb11efb9bb066c4d81d0bab
<pre>
[GPU Process] [FormControls] Make it possible to draw a viewless NSCell
<a href="https://bugs.webkit.org/show_bug.cgi?id=252003">https://bugs.webkit.org/show_bug.cgi?id=252003</a>
rdar://105233927

Reviewed by Aditya Keerthi.

AppKit crashes if [NSCell drawWithFrame] is called out of the main thread with a
non-nil NSView. And this is why a hack was added to WebControlWindow to silence
this crash in 257981@main. To fix the crash and to remove the hack, a nil NSView
will be passed to [NSCell drawWithFrame] instead of passing the fake one.

To provide the [NSCell drawWithFrame] with the settings that it used to get from
the fake NSView, new AppKit SPIs will be used to explicitly set them from the
drawing. An NSCell can be drawn without an NSView only if these SPIs are available.

* Source/WebCore/PAL/pal/spi/mac/NSCellSPI.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::supportsViewlessCells):
(WebCore::shouldFlipContext):
(WebCore::applyViewlessCellSettings):
(WebCore::drawViewlessCell):
(WebCore::drawCellInView):
(WebCore::drawCellFocusRingInView):
(WebCore::drawCellFocusRing):
(WebCore::ControlMac::drawCellOrFocusRing):
(WebCore::ControlMac::drawCell):
(WebCore::drawFocusRing): Deleted.
(WebCore::drawCellOrFocusRing): Deleted.
* Source/WebCore/platform/graphics/mac/controls/WebControlView.mm:

Canonical link: <a href="https://commits.webkit.org/260210@main">https://commits.webkit.org/260210@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9bacde32b4b9715ab7a1d61657461c7c11d6b103

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107049 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116225 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115735 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110950 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7298 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99226 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13276 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96309 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40883 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27948 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82634 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9205 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29299 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9802 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6303 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15364 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48863 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7059 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11344 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -211,8 +249,15 @@ static void drawCellOrFocusRing(GraphicsContext& context, const FloatRect& rect,
-            [cell drawWithFrame:rect inView:view];
-            [cell setControlView:nil];
+            // Check if we can draw the NSCell without the fake NSView.
+            if ([NSCell instancesRespondToSelector:@selector(_setFallbackBackingScaleFactor:)]) {
```
could have been `[cell respondsToSelector:...]` instead. Also you can cache it.

hortont424 replied:
And maybe factor it out into a static? "supportsViewlessCells()" or something, so that it's not so mysterious why you're checking something with a wildly different name

shallawa replied:
I added the static `supportsViewlessCells()` and cached the value so I have to use `[NSCell instancesRespondToSelector:...]` but just for a single time.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -176,6 +177,43 @@
+    [cell setControlView:nil];
+}
+
 static void drawFocusRing(NSRect cellFrame, NSCell *cell, NSView *view)
```
Do we need a viewless approach here as well?

shallawa replied:
Yes we do. But when passing a nil NSView to [cell drawFocusRingMaskWithFrame:...] nothing is displayed. I added a FIXME comment and I filed rdar://105249508.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -176,6 +177,43 @@
+    }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wnonnull"
```
We should probably ask AppKit for a version of this method which that doesn't require a view, so that we don't need these.

shallawa replied:
I added a FIXME comment and I filed rdar://105250010.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/WebControlView.mm:
```diff
@@ -60,6 +60,8 @@ - (BOOL)isKeyWindow
     return _hasKeyAppearance;
 }
 
+// FIXME: These are hacks to prevent an AppKit crash when drawing an NSCell in GPUP
```
Is there a radar you could leave here? Also, missing a period at the end of the first sentence.

shallawa replied:
I modified the FIXME comment and I filed rdar://105250168.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -211,8 +249,15 @@ static void drawCellOrFocusRing(GraphicsContext& context, const FloatRect& rect,
+            // Check if we can draw the NSCell without the fake NSView.
+            if ([NSCell instancesRespondToSelector:@selector(_setFallbackBackingScaleFactor:)]) {
+                // Don't flip the context if we are just drawing an image.
+                bool flipContext = m_owningPart.type() != StyleAppearance::Checkbox
```
Can this logic also be removed once AppKit has fixed rdar://105194487?

Also, it would be nice to put this in a static method.

shallawa replied:
I moved it to a static method called `shouldFlipContext()`.

shallawa committed:
[GPU Process] [FormControls] Make it possible to draw a viewless NSCell
https://bugs.webkit.org/show_bug.cgi?id=252003
rdar://105233927

Reviewed by Aditya Keerthi.

AppKit crashes if [NSCell drawWithFrame] is called out of the main thread with a
non-nil NSView. And this is why a hack was added to WebControlWindow to silence
this crash in 257981@main. To fix the crash and to remove the hack, a nil NSView
will be passed to [NSCell drawWithFrame] instead of passing the fake one.

To provide the [NSCell drawWithFrame] with the settings that it used to get from
the fake NSView, new AppKit SPIs will be used to explicitly set them from the
drawing. An NSCell can be drawn without an NSView only if these SPIs are available.

* Source/WebCore/PAL/pal/spi/mac/NSCellSPI.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::supportsViewlessCells):
(WebCore::shouldFlipContext):
(WebCore::applyViewlessCellSettings):
(WebCore::drawViewlessCell):
(WebCore::drawCellInView):
(WebCore::drawCellFocusRingInView):
(WebCore::drawCellFocusRing):
(WebCore::ControlMac::drawCellOrFocusRing):
(WebCore::ControlMac::drawCell):
(WebCore::drawFocusRing): Deleted.
(WebCore::drawCellOrFocusRing): Deleted.
* Source/WebCore/platform/graphics/mac/controls/WebControlView.mm:

Canonical link: https://commits.webkit.org/260210@main
